hail/src/main/scala/is/hail/types/physical/stypes/interfaces/SNDArray.scala

```diff
@@ -41,6 +44,104 @@ object SNDArray {
+    if (arrays.isEmpty) return
+    val indexVars = Array.tabulate(arrays(0)._1.st.nDims)(i => s"i$i").toFastIndexedSeq
+    val indices = Array.range(0, arrays(0)._1.st.nDims).toFastIndexedSeq
+    coiterate(cb, region, indexVars, arrays.map { case (array, name) => (array, indices, name) }, body, deepCopy)
+  }
+
+  def coiterate(cb: EmitCodeBuilder, region: Value[Region], indexVars: IndexedSeq[String], arrays: IndexedSeq[(SNDArrayCode, IndexedSeq[Int], String)], body: IndexedSeq[SSettable] => Unit): Unit =
+    coiterate(cb, region, indexVars, arrays, body, deepCopy=false)
+
+  // Note: to iterate through an array in column major order, make sure the indices are in ascending order. E.g.
+  // coiterate(cb, region, IndexedSeq("i", "j"), IndexedSeq((A, IndexedSeq(0, 1), "A"), (B, IndexedSeq(0, 1), "B")), {
+  //   case Seq(a, b) => cb.assign(a, SCode.add(cb, a, b))
+  // })
+  // computes A += B.
+  def coiterate(cb: EmitCodeBuilder, region: Value[Region], indexVars: IndexedSeq[String], arrays: IndexedSeq[(SNDArrayCode, IndexedSeq[Int], String)], body: IndexedSeq[SSettable] => Unit, deepCopy: Boolean): Unit = {
+
+    val indexSizes = new Array[Settable[Int]](indexVars.length)
+    val indexCoords = Array.tabulate(indexVars.length) { i => cb.newLocal[Int](indexVars(i)) }
+
+    case class ArrayInfo(
+      array: SNDArrayValue,
+      strides: IndexedSeq[Value[Long]],
+      pos: IndexedSeq[Settable[Long]],
+      elt: SSettable,
+      indexToDim: Map[Int, Int],
+      name: String)
+
+    val info = arrays.map { case (_array, indices, name) =>
+      for (idx <- indices) assert(idx < indexVars.length && idx >= 0)
+      // FIXME: relax this assumption to handle transposing, non-column major
+      for (i <- 0 until indices.length - 1) assert(indices(i) < indices(i+1))
+      assert(indices.length == _array.st.nDims)
+
+      val array = _array.memoize(cb, s"${name}_copy")
+      val shape = array.shapes(cb)
+      for (i <- indices.indices) {
+        val idx = indices(i)
+        if (indexSizes(idx) == null) {
+          indexSizes(idx) = cb.newLocal[Int](s"${indexVars(idx)}_max")
+          cb.assign(indexSizes(idx), shape(i).toI)
+        } else {
+          cb.ifx(indexSizes(idx).cne(shape(i).toI), s"${indexVars(idx)} indexes incompatible dimensions")
+        }
+      }
+      val strides = array.strides(cb)
+      val pos = Array.tabulate(array.st.nDims + 1) { i => cb.newLocal[Long](s"$name$i") }
+      val elt = new SSettable {
+        def st: SType = array.st.elementType
+        val pt: PType = array.st.pType.elementType
+
+        // FIXME: need to use `pos` of smallest index var
+        def get: SCode = pt.loadCheapPCode(cb, pt.loadFromNested(pos(0)))
+        def store(cb: EmitCodeBuilder, v: SCode): Unit = pt.storeAtAddress(cb, pos(0), region, v, deepCopy)
+        def settableTuple(): IndexedSeq[Settable[_]] = FastIndexedSeq(pos.last)
+      }
+      val indexToDim = Map(indices.indices.map(i => indices(i) -> i): _*)
```

Comment:
This is `indices.zipWithIndex`, isn't it?

Reply:
Like, `indexToDim = Map(indices.zipWithIndex(): _*)`

Reply:
Ah, yep. I'll change it.